### PR TITLE
Ensure `acos(h)`/`asin(h)` respect special cases from applicable standards

### PR DESCRIPTION
**Pull request by mdhaber**
As discussed in gh-1172, this PR resolves the disagreement between `acos`'s current behavior and that specified in the [array API standard](https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acos.html).

Because it is my first contribution, this PR is intended to be minimal and follow nearby patterns as closely as possible. Let me know what you'd prefer to see in the future and I'll make it so.

A more complete version, which tests _all_ the special cases from the standard - not just the ones that were failing - can be seen at:
https://github.com/mdhaber/mpmath/pull/new/fix_all_acos_special_cases
I'd prefer to go that route with this and future PRs (for other trig, hyperbolic trig, etc.), but it is more to review, so I decided to submit the minimal version. Let me know which direction you'd prefer. I've marked this as draft in the meantime.

When working on SciPy or MPArray, I never let AI draft my code. I did here because the changes are very mechanical. Of course, it (GPT 5.6 Sol, medium effort) made many mistakes, but I think I fixed it all up so it is compact and consistent with the existing code.

### CODE REVIEW

File: mpmath/libmp/libmpc.py [hunk 1]
```diff
@@ -767,6 +767,11 @@ def acos_asin(z, prec, rnd, n):
     return re, im
 
 def mpc_acos(z, prec, rnd=round_down):
+    a, b = z
+    if a == fzero and b == fnan:
+        return mpf_shift(mpf_pi(prec, rnd), -1), fnan
+    if a == fnan and b not in _infs:
+        return fnan, fnan
     return acos_asin(z, prec, rnd, 0)
```
Comment (skirpichev) [inline mpmath/libmp/libmpc.py:775]: I suspect, that the right place to put the fix is acos_asin() helper.

File: mpmath/tests/test_functions.py [hunk 2]
```diff
@@ -464,6 +464,16 @@ def test_acos():
     assert acos(mpc(+2, 0)).ae(mpc(0, log(2 + sqrt(3))))
     assert acos(mpc(0.5, 0)).ae(pi/3)
 
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acos.html
+    # "If a is either +0 or -0 and b is NaN": pi/2 + NaN j.
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py:469]:
```suggestion
```

I think no need to comment such things.

File: mpmath/tests/test_functions.py [hunk 3]
```diff
@@ -464,6 +464,16 @@ def test_acos():
...
+    r = acos(mpc(0, nan))
+    assert r.real.ae(pi/2) and isnan(r.imag)
+
+    # "If a is NaN and b is a finite number": NaN + NaN j.
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py:473]:
```suggestion
```

**Review by skirpichev — COMMENTED** [review]
> I decided to submit the minimal version. Let me know which direction you'd prefer.

Please submit all remaining tests, but avoid much rewriting as you did in the mentioned above branch.

File: mpmath/tests/test_functions.py (hunk 2 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py]: Ok, I thought that would make them easier to verify.  But I'll remove them and just order the assertions as they appear in the standard; they should still be easy enough to compare.

I do think I should keep the comment that indicates the source of the special case tests though. That's just one line - the URL of the function in the standard.

File: mpmath/libmp/libmpc.py (hunk 1 above)
Comment (mdhaber) [reply mpmath/libmp/libmpc.py:776]: I moved one of the cases there because it is the same for both `acos` and `asin`, but I left this one here because it is different for the two functions.

File: mpmath/tests/test_functions.py [hunk 4]
```diff
@@ -435,26 +435,17 @@ def test_asin():
 
 def test_acos():
     pi4 = pi/4
-    assert acos(mpc(+inf, +inf)) == mpc(+pi4, -inf)
```
Comment (mdhaber) [inline mpmath/tests/test_functions.py:438]: I'm torn about whether to remove these. It is logical to keep them together, but I don't want to be redundant. Let me know if you'd prefer for them to be left in and be a little redundant.

(I'd really rather not try to merge the array API standard cases into these because I think the standard's cases should be in order and intact.)

File: mpmath/tests/test_functions.py (hunk 2 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py]: Removed.

File: mpmath/tests/test_functions.py (hunk 3 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py]: Done.

File: mpmath/libmp/libmpc.py (hunk 1 above)
Comment (skirpichev) [reply mpmath/libmp/libmpc.py:776]: Yes, this is different, but asin() is broken just as acos for this input:
```pycon
>>> mp.asin(complex('0+nanj'))
(nan + nanj)
>>> fp.asin(complex('0+nanj'))
nanj
>>> fp.asin(complex('-0.0+nanj'))
(-0+nanj)
```

I would prefer to fix *both* functions.  But in any case: all logic belongs to the acos_asin(), not to the wrapper.

File: mpmath/libmp/libmpc.py [hunk 5]
```diff
@@ -621,6 +621,8 @@ def acos_asin(z, prec, rnd, n):
     and by abs(a) <= 1, in order to improve the numerical accuracy.
     """
     a, b = z
+    if a == fnan and b not in _infs:
+        return fnan, fnan
```
Comment (skirpichev) [inline mpmath/libmp/libmpc.py]: I think that major part of nan's input will be handled correctly.  Lets make a more targeted fix.

I believe, you should correct L633/635 below.  Something like this:
```diff
diff --git a/mpmath/libmp/libmpc.py b/mpmath/libmp/libmpc.py
index 23bbacef..88eb0771 100644
--- a/mpmath/libmp/libmpc.py
+++ b/mpmath/libmp/libmpc.py
@@ -630,7 +630,7 @@ def acos_asin(z, prec, rnd, n):
         # case abs(a) <= 1
         if not am[0]:
             if n == 0:
-                return mpf_acos(a, prec, rnd), fzero
+                return mpf_acos(a, prec, rnd), fzero if a != fnan else a
             else:
                 return mpf_asin(a, prec, rnd), fzero
         # cases abs(a) > 1
```
?

File: mpmath/tests/test_functions.py (hunk 4 above)
Comment (skirpichev) [reply mpmath/tests/test_functions.py:438]: Please don't remove tests (unless they are really duplicates).

> I'd really rather not try to merge the array API standard cases into these because I think the standard's cases should be in order and intact.

I understand, that tests could be written better, cases represented in a more clear order and so on.  But such refactoring just for one test function looks like code churn for me, sorry.

Please just add missing tests.

File: mpmath/tests/test_functions.py [hunk 6]
```diff
@@ -464,6 +455,33 @@ def test_acos():
     assert acos(mpc(+2, 0)).ae(mpc(0, log(2 + sqrt(3))))
     assert acos(mpc(0.5, 0)).ae(pi/3)
 
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acos.html
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]: I would prefer be compatible with the C standard.  Are there any differences?

File: mpmath/libmp/libmpc.py (hunk 5 above)
Comment (mdhaber) [reply mpmath/libmp/libmpc.py]: I suppose I was optimizing for the most obviously correct fix when I put it in the `acos` definition. I typically try to avoid quadruply nested conditions, but I can adopt this and try to find more clever fixes in the future.

**Comment by mdhaber** [comment]
> Yes, this is different, but asin() is broken just as acos for this input:

The plan was to fix `asin` and add those tests in a separate PR. Personally, I prefer to review targeted PRs, but I can create a mega PR with all fixes at once if you prefer.

File: mpmath/tests/test_functions.py (hunk 6 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py]: I don't think there are any incompatibilities between the standards. The special cases appear to have one to one correspondence, and they are listed in the same order.
[Array API `acos`](https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acos.html)
[C `cacosf`](https://en.cppreference.com/c/numeric/complex/cacos)
[C++ `std::acos(std::complex)`](https://en.cppreference.com/cpp/numeric/complex/acos)
I can check whether there are differences for the other functions - I doubt they strayed from existing standards unless there was a very good reason.

I think the only incompatibility is that `mpmath` gives complex analytic continuation for real input whereas the standards reserve analytic continuation for complex input, but I am not proposing to change that.

File: mpmath/tests/test_functions.py (hunk 4 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py:438]: They are exact duplicates.

> looks like code churn for me, sorry.

Right, I was saying that I'd prefer not to do that.

IIUC, you want me to leave the existing tests alone and to add only the cases from the standard that are not already checked. I hesitated to do that because it will be more difficult for you to see that the tests are complete, but I will follow the instruction.

File: mpmath/tests/test_functions.py (hunk 6 above)
Comment (skirpichev) [reply mpmath/tests/test_functions.py]: > I think the only incompatibility is that mpmath gives complex analytic continuation for real input whereas the standards reserve analytic continuation for complex input, but I am not proposing to change that.

mpmath also can raise an exception, if mp.trap_complex is enabled, e.g.:
```pycon
>>> mp.trap_complex = True
>>> acos(2)
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    acos(2)
    ~~~~^^^
  File "/home/sk/src/mpmath/mpmath/ctx_mp_python.py", line 1188, in f
    return ctx.make_mpf(mpf_f(x._mpf_, prec, rounding))
                        ~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/sk/src/mpmath/mpmath/libmp/libelefun.py", line 953, in mpf_acos
    raise ComplexResult("acos(x) is real only for -1 <= x <= 1")
mpmath.libmp.libmpf.ComplexResult: acos(x) is real only for -1 <= x <= 1
```

**Comment by skirpichev** [comment]
> The plan was to fix asin and add those tests in a separate PR.

Up to you, but this doesn't look as a mega-pr (an example: https://github.com/mpmath/mpmath/pull/1166) for me.

File: mpmath/libmp/libmpc.py (hunk 5 above)
Comment (skirpichev) [reply mpmath/libmp/libmpc.py]: > I typically try to avoid quadruply nested conditions

This is rather niche case (I doubt that nans input is "normal").  I would prefer to avoid special cases rather than introduce them.

File: mpmath/libmp/libmpc.py (hunk 1 above)
Comment (mdhaber) [reply mpmath/libmp/libmpc.py:776]: Moved the logic into `acos_asin`, preferring modifications that are less likely to require additional condition evaluations.

File: mpmath/tests/test_functions.py (hunk 4 above)
Comment (mdhaber) [reply mpmath/tests/test_functions.py:438]: Left the existing tests alone and added only additional tests needed to test the cases required by the standard.

**Comment by mdhaber** [comment]
OK, added the conditions in `acos_asin`, and tried to put them where they were less likely to add computation, since they are edge cases rather than typical input. Didn't alter any existing tests, just added (only) missing special case tests based on the standard. With the few targeted changes, all the tests are passing now for `acos`, `asin`, `acosh`, and `asinh`.

File: mpmath/tests/test_functions.py [hunk 7]
```diff
@@ -433,6 +433,21 @@ def test_asin():
     assert asin(mpc(0, 1e-220)).ae(1e-220j)
     mp.prec = 53
 
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.asin.html
+    assert isnan(asin(mpf(nan)))
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert isnan(asin(nan))
```

File: mpmath/tests/test_functions.py [hunk 8]
```diff
@@ -433,6 +433,21 @@ def test_asin():
...
+    assert asin(mpf(0)) == 0
+    assert asin(mpc(0, 0)) == 0
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert asin(0) == 0
    assert asin(0j) == 0
```

File: mpmath/tests/test_functions.py [hunk 9]
```diff
@@ -464,6 +479,80 @@ def test_acos():
     assert acos(mpc(+2, 0)).ae(mpc(0, log(2 + sqrt(3))))
     assert acos(mpc(0.5, 0)).ae(pi/3)
 
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acos.html
+    assert isnan(acos(mpf(nan)))
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert isnan(acos(nan))
```

File: mpmath/tests/test_functions.py [hunk 10]
```diff
@@ -464,6 +479,80 @@ def test_acos():
...
+    assert acos(mpf(1)).ae(0)
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert acos(1) == 0
```

File: mpmath/tests/test_functions.py [hunk 11]
```diff
@@ -464,6 +479,80 @@ def test_acos():
...
+    assert acos(mpc(0, 0)).ae(mpc(pi2, 0))
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert acos(0j).ae(mpc(pi2, 0))
```

File: mpmath/tests/test_functions.py [hunk 12]
```diff
@@ -464,6 +479,80 @@ def test_acos():
...
+    pi2 = pi/2
+    pi4 = pi/4
+
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.asinh.html
+    assert isnan(asinh(mpf(nan)))
+    assert asinh(mpf(0)) == 0
+    assert asinh(mpf(inf)) == inf
+    assert asinh(mpf(-inf)) == -inf
+    assert asinh(mpc(0, 0)) == 0
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert isnan(asinh(nan))
    assert asinh(0) == 0
    assert asinh(inf) == inf
    assert asinh(-inf) == -inf
    assert asinh(0j) == 0
```
Lets use literals or mpmath's constants, where possible.

File: mpmath/tests/test_functions.py [hunk 13]
```diff
@@ -464,6 +479,80 @@ def test_acos():
...
+
+def test_acosh():
+    pi2 = pi/2
+    pi4 = pi/4
+
+    # Special cases:
+    # https://data-apis.org/array-api/2025.12/API_specification/generated/array_api.acosh.html
+    assert isnan(acosh(mpf(nan)))
+    assert acosh(mpf(1)) == 0
+    assert acosh(mpf(inf)) == inf
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert isnan(acosh(nan))
    assert acosh(1) == 0
    assert acosh(inf) == inf
```

File: mpmath/tests/test_functions.py [hunk 14]
```diff
@@ -464,6 +479,80 @@ def test_acos():
...
+    assert acosh(mpc(0, 0)).ae(mpc(0, pi2))
```
Comment (skirpichev) [inline mpmath/tests/test_functions.py]:
```suggestion
    assert acosh(0j).ae(mpc(0, pi2))
```

File: mpmath/libmp/libmpc.py [hunk 15]
```diff
@@ -740,9 +740,11 @@ def acos_asin(z, prec, rnd, n):
     if im[3] >= 0:
         im = normalize(im[0], im[1], im[2], im[3], prec, rnd)
     # Correct real part for infinities and nan in imaginary component
-    if re == fnan and mpc_is_inf(z):
+    if re == fnan and (mpc_is_inf(z) or z == (fzero, fnan)):
```
Comment (skirpichev) [inline mpmath/libmp/libmpc.py]: Rather you could add specialization for pure-imaginary arguments (z.real == fzero), just like for pure-real case.  For now - only that required for your tests.

File: mpmath/tests/test_functions.py (hunk 6 above)
Comment (skirpichev) [reply mpmath/tests/test_functions.py]: I would prefer to quote the C standard.

File: mpmath/libmp/libmpc.py (hunk 15 above)
Comment (mdhaber) [reply mpmath/libmp/libmpc.py]: It sounded like you wanted the special cases to be evaluated less often since they are not typical input, so I buried them. But yes, a top-level pure imaginary branch is also possible - done in the next commit.

**Comment by mdhaber** [comment]
Now quoting the C standard and using literals or constants where possible.

**Review by skirpichev — APPROVED** [review]
Let me know if you are ok with squashing this in one commit.  Then please adjust pr title.  I also appreciate if you provide some draft for commit message.

Or rebase pr to several commits, if you prefer.

Thanks for patience.

**Comment by mdhaber** [comment]
Squash merge is fine. 

Commit message could be: 
Add special case logic and regression tests to ensure that `acos`, `asin`, `acosh`, and `asinh` respect special cases defined in the C and Python array API standards.

Written with support from GPT 5.6 Sol

You are also welcome to credit yourself with a "Co-authored by:" note, even though I didn't hit apply on your suggestions directly. Please adjust either the title or commit message as you prefer, or ask for adjustments if needed. 

Thanks for the review!

**Comment by skirpichev** [comment]
Thanks.
